servers/ramatak/ad_server.cpp

```diff
@@ -0,0 +1,200 @@
+		Ref<AdPlugin> plugin = this->ad_plugins[plugin_name];
+		if (plugin.is_null()) {
+			WARN_PRINT_ONCE(vformat("Plugin is null during initialization: %s", plugin_name));
+			continue;
+		}
+		AdPlugin::AdPluginResult result = plugin->init_plugin();
+		switch (result) {
+			case AdPlugin::AdPluginResult::OK:
+				break;
+			case AdPlugin::AdPluginResult::PLUGIN_NOT_ENABLED_WARN:
+				WARN_PRINT(vformat("Plugin not enabled, but requested: %s", plugin_name));
+				return;
+			case AdPlugin::AdPluginResult::AUTH_ERROR:
+			case AdPlugin::AdPluginResult::NETWORK_ERROR:
+			case AdPlugin::AdPluginResult::UNSPECIFIED_ERROR:
+				WARN_PRINT(vformat("Plugin initialization failed: %s", plugin_name));
+				break;
+		}
+	}
+}
+
+AdServer *AdServer::get_singleton() {
+	return AdServer::singleton;
+}
+
+Array AdServer::get_available_plugins() const {
+	return this->ad_plugin_list;
+}
+
+Ref<AdPlugin> AdServer::get_plugin_raw(String p_name) {
+	return this->ad_plugins.get(p_name);
+}
+
+Array AdServer::get_plugin_priority_order() const {
+	return ProjectSettings::get_singleton()->get("ramatak/monetization/ad_plugin_priorities");
+}
+
+void AdServer::set_plugin_priority_order(Array p_priorites) {
+	ProjectSettings::get_singleton()->set("ramatak/monetization/ad_plugin_priorities", p_priorites);
+	ProjectSettings::get_singleton()->save();
+}
+
+Dictionary AdServer::get_plugin_config(String p_plugin) const {
+	Dictionary all_configs = ProjectSettings::get_singleton()->get("ramatak/monetization/ad_plugin_config");
+	return all_configs[p_plugin];
+}
+
+Variant AdServer::show_other(String p_ad_unit) {
+	Array ad_plugin_priorities = (Array)ProjectSettings::get_singleton()->get("ramatak/monetization/ad_plugin_priorities");
+	String ad_unit_setting = "ramatak/monetization/ad_units";
+	ProjectSettings *project_settings = ProjectSettings::get_singleton();
+
+	Dictionary dict = project_settings->get(ad_unit_setting);
+	Dictionary ad_unit_config = dict[p_ad_unit];
+	int ad_type = (int)ad_unit_config["ad_unit_type"];
+
+	for (int i = 0; i < ad_plugin_priorities.size(); i++) {
+		String network_specific_id = MonetizationSettings::get_singleton()->get_ad_unit_network_id(p_ad_unit, ad_plugin_priorities[i]);
+		if (network_specific_id.length() > 0) {
+			// Plugin `i` can handle the specified ad_unit.
+			Variant request_token = rand.rand() & 0x7FFFFFFF;
+			ad_plugins[ad_plugin_priorities[i]]->show_other(network_specific_id, request_token, (AdServer::AdType)ad_type);
+			return request_token;
+		}
+	}
+	WARN_PRINT_ONCE("No ad plugin is capable of handling ad unit.");
+	// There might be a better return value here.
+	return Variant();
+}
+
+Variant AdServer::show_banner(String p_ad_unit, BannerAdSize p_size, BannerAdLocation p_location) {
+	// This looks gross, and it is, but these values are going to cross an FFI boundary into a plugin and
+	// it's much less error-prone to pass a string than an integer, especially because these plugins might
```

Comment:
TODO: clarify that I mean an integer representation of an unstable enum that might change from one minor release to the next.